task_003a.rb

```diff
@@ -0,0 +1,152 @@
+#     Then, consider the second bit of the 5 remaining numbers: there are fewer 1 bits (2) than 0 bits (3), so keep
+#     only the 2 numbers with a 1 in the second position: 01111 and 01010.
+#     In the third position, there are an equal number of 0 bits and 1 bits (one each). So, to find the CO2 scrubber
+#     rating, keep the number with a 0 in that position: 01010.
+#     As there is only one number left, stop; the CO2 scrubber rating is 01010, or 10 in decimal.
+#
+# Finally, to find the life support rating, multiply the oxygen generator rating (23) by the CO2 scrubber rating (10)
+# to get 230.
+#
+# Use the binary numbers in your diagnostic report to calculate the oxygen generator rating and CO2 scrubber rating,
+# then multiply them together. What is the life support rating of the submarine? (Be sure to represent your
+# answer in decimal, not binary.)
+
+def parse_input(input_data)
+  binary_diagnostic_report = []
+
+  input_data.split("\n").each do |binary_value|
+    binary_diagnostic_report << binary_value.each_char.to_a
+  end
+  binary_diagnostic_report
+end
+
+def find_common(binary_diagnostic_report, bit_position)
+  most_common = 0
+
+  binary_diagnostic_report.each do |value|
+    case value[bit_position].to_i
+    when 1 then most_common += 1
+    when 0 then most_common -= 1
+    end
+  end
+  most_common
+end
+
+def bit_criteria(binary_diagnostic_report, common) # common can be: "most" or "least"
+  bit_position = 0
+
+  # binary_report = binary_diagnostic_report.dup  # can use this instead lines below
+
+  # code from this line ...
+
+  most_common = find_common(binary_diagnostic_report, bit_position)
+
+  case common
+  when "most" then
+    if most_common >= 0
+      binary_report = binary_diagnostic_report.reject {|value| value[bit_position] == "0"}
+    else
+      binary_report = binary_diagnostic_report.reject {|value| value[bit_position] == "1"}
+    end
+
+  when "least" then
+    if most_common >= 0
+      binary_report = binary_diagnostic_report.reject {|value| value[bit_position] == "1"}
+    else
+      binary_report = binary_diagnostic_report.reject {|value| value[bit_position] == "0"}
+    end
+  end
+
+  bit_position += 1     # ... to here, is needed to not lose: binary_diagnostic report
+
+  while (binary_report.size > 1 && bit_position < binary_diagnostic_report[0].size) do
+    most_common = find_common(binary_report, bit_position)
+
+    case common
+    when "most" then
+      if most_common >= 0
+        binary_report.delete_if {|value| value[bit_position] == "0"} #delete_if can be: reject!
+      else
+        binary_report.delete_if {|value| value[bit_position] == "1"}
+      end
+
+    when "least" then
+      if most_common >= 0
+        binary_report.delete_if {|value| value[bit_position] == "1"}
+      else
+        binary_report.delete_if {|value| value[bit_position] == "0"}
+      end
+    end
+
+    bit_position += 1
+  end
+
+  binary_report[0].join
```

Comment:
One solution was to use `binary_report = binary_diagnostic_report.dup` and then `delete_if` as you tried.

The other solution would be to avoid `binary_diagnostic_report.dup` and `delete_if` completely:

```ruby
binary_diagnostic_report[0].size.times do |bit_position|
  break if binary_diagnostic_report.size == 1

  most_common = find_common(binary_diagnostic_report, bit_position)

  binary_diagnostic_report = case common
    when 'most'
      if most_common >= 0
        binary_diagnostic_report.reject { |binary_numbers| binary_numbers[bit_position] == '0' }
      else
        binary_diagnostic_report.reject { |binary_numbers| binary_numbers[bit_position] == '1' }
      end

    when 'least'
      if most_common >= 0
        binary_diagnostic_report.reject { |binary_numbers| binary_numbers[bit_position] == '1' }
      else
        binary_diagnostic_report.reject { |binary_numbers| binary_numbers[bit_position] == '0' }
      end
    end
end

binary_diagnostic_report[0].join
```